### PR TITLE
Address backwards compatibility issue with AssemblyInformationalVersionAttribute handling

### PR DESCRIPTION
**Pull request by Giftednewt**
Make compiler produce the same Product Build, Major, Minor and Private Parts in VersionInfo as VS2013 compiler when Assembly ProductAssemblyFileVersionAttribute and AssemblyInformationalVersion contain invalid format.

Also introduce warning diagnostic for AssemblyInformationalVersion with invalid version string.

This addresses issue #5098

### CODE REVIEW

**Comment by Giftednewt** [comment]
Tagging @gafter

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Please review this external submission for 1.3.

**Comment by davkean** [comment]
@dotnet-bot test eta please

File: src/Compilers/Core/Portable/VersionHelper.cs [hunk 1]
```diff
@@ -65,18 +66,43 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
...
             for (int i = 0; i < lastExplicitValue; i++)
             {
-                if (!ushort.TryParse(elements[i], NumberStyles.None, CultureInfo.InvariantCulture, out values[i]) || values[i] > maxValue)
+                
+                if (allowPartialParse)
                 {
-                    version = null;
-                    return false;
+                    //Shorten the version string until we are at the shortest possible string that can be represented as a ushort
+                    while (!ushort.TryParse(elements[i], NumberStyles.None, CultureInfo.InvariantCulture, out values[i]))
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/VersionHelper.cs]: It looks like on the `allowPartialParse` path we are no longer doing the `values[i] > maxValue` check.  I believe that check should still be there.

**Comment by AlekseyTs** [comment]
CC @tmat

File: src/Compilers/Core/CodeAnalysisTest/VersionHelperTests.cs [hunk 2]
```diff
@@ -88,95 +88,104 @@ public void ParseGood2()
...
             Assert.False(VersionHelper.TryParse("1.2. 3", out version));
-            Assert.Null(version);
+            Assert.Equal(new Version(1, 2, 0, 0), version);
             Assert.False(VersionHelper.TryParse("1.2.3 ", out version));
-            Assert.Null(version);
+            Assert.Equal(new Version(1, 2, 3, 0), version);
             Assert.False(VersionHelper.TryParse("1.a", out version));
-            Assert.Null(version);
+            Assert.Equal(new Version(1, 0, 0, 0), version);
             Assert.False(VersionHelper.TryParse("1.2.a.b", out version));
```
Comment (tmat) [inline src/Compilers/Core/CodeAnalysisTest/VersionHelperTests.cs]: Missing cases with large numbers, empty parts:
1.1.65536, 1.1.1.10000000, 1..1.2

File: src/Compilers/Core/Portable/VersionHelper.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/Core/Portable/VersionHelper.cs]: I think we should rather find the next non-digit (0-9) and then parse it just once.

**Comment by Giftednewt** [comment]
@tmat I've altered the parsing so that it finds the first non digit and then parses the string up to that point

@jaredpar I looked at the behavior of the old compiler and it appears that when the element value is over 65535 it uses the value of the 16 least significant bits and returns that value for the element part. I've introduced this into the allowPartialParse code path

**Comment by tmat** [comment]
BTW, you can use http://source.roslyn.io/#Microsoft.CodeAnalysis/MetadataReference/AssemblyIdentity.cs,242 instead of allocating a new one each time you need to return 0.0.0.0

**Comment by Giftednewt** [comment]
Cool. I'll get that changed over when I get a chance

**Comment by gafter** [comment]
LGTM

@dotnet/roslyn-compiler Please review this external submission for 1.3.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Please review this external submission for 1.3.

**Comment by gafter** [comment]
@tmat We need a second review approval for this PR. Do you think it is in good shape now?

**Comment by tmat** [comment]
:+1:

**Comment by gafter** [comment]
@Giftednewt THANK YOU THANK YOU THANK YOU for your contribution!

**Comment by Giftednewt** [comment]
@gafter No problem, glad to have been able to help

File: src/Compilers/Core/Portable/VersionHelper.cs [hunk 3]
```diff
@@ -65,21 +66,77 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
...
+                    if(values[i] > maxValue)
+                    {
+                        //The only way this can happen is if the value was 65536
+                        //The old compiler would continue parsing from here
+                        values[i] = 0;
+                        continue;
+                    }
+
+                    bool invalidFormat = false;
+                    System.Numerics.BigInteger number = 0;
```
Comment (cston) [inline src/Compilers/Core/Portable/VersionHelper.cs]: `number` is not used.

**Pull request merged**